[Site isolation] Record the page snapshotting to a DisplayList before sinking it to a PDFDocument

shallawa opened the pull request:
#### fc01b4369ae0acd8cdda47914bd506df0a6c6f9d
<pre>
[Site isolation] Record the page snapshotting to a DisplayList before sinking it to a PDFDocument
<a href="https://bugs.webkit.org/show_bug.cgi?id=285726#">https://bugs.webkit.org/show_bug.cgi?id=285726#</a>
<a href="https://rdar.apple.com/142664977">rdar://142664977</a>

Reviewed by Matt Woodrow.

Introduce ImageBufferDisplayListBackend and ImageBufferRemoteDisplayListBackend
to support recording the snapshot to a DisplayList.

Make drawRemoteToPDF() creates a ImageBufferRemoteDisplayListBackend. Its GPUProcess
twin ImageBufferDisplayListBackend will record the drawing commands into a
DisplayList.

Make didDrawRemoteToPDF() calls sinkIntoPDFDocument() on the main thread because
DisplayList::Replayer needs a ControlFactory and ControlFactory::shared() has to
be called on the main thread. Then send the result SharedBuffer to the UIProcess.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::sinkIntoPDFDocument):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferDisplayListBackend.cpp: Added.
(WebCore::ImageBufferDisplayListBackend::create):
(WebCore::ImageBufferDisplayListBackend::ImageBufferDisplayListBackend):
(WebCore::ImageBufferDisplayListBackend::context):
(WebCore::ImageBufferDisplayListBackend::copyNativeImage):
(WebCore::ImageBufferDisplayListBackend::sinkIntoPDFDocument):
(WebCore::ImageBufferDisplayListBackend::debugDescription const):
* Source/WebCore/platform/graphics/ImageBufferDisplayListBackend.h: Copied from Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp.
(WebCore::ImageBufferDisplayListBackend::calculateMemoryCost):
* Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp:
(WebCore::DisplayList::DrawingContext::replayDisplayList):
The PDFDocument has to be generated on the main thread because the main frame
DisplayList will eventually reference other DisplayLists which were created by
different WorkQueues. So clearing the DisplayList, once it is replayed back, will
delete the resources in the ResourceHeap on the main thread. This will fire the
assertion: &quot;Unsafe to ref/deref from different threads&quot;. So keep the DisplayList
items till they are deleted by the destructor of DisplayList::DrawingContext.

* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didDrawRemoteToPDF):
(WebKit::allocateImageBufferInternal):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemoteDisplayListBackend.cpp: Added.
(WebKit::ImageBufferRemoteDisplayListBackend::create):
(WebKit::ImageBufferRemoteDisplayListBackend::ImageBufferRemoteDisplayListBackend):
(WebKit::ImageBufferRemoteDisplayListBackend::~ImageBufferRemoteDisplayListBackend):
(WebKit::ImageBufferRemoteDisplayListBackend::createNativeImageReference):
(WebKit::ImageBufferRemoteDisplayListBackend::debugDescription const):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemoteDisplayListBackend.h: Copied from Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):

(WebKit::RemoteImageBufferProxy::ensureBackend const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::sendSync):
Fix unrelated mac-safer-cpp errors. Make these functions make early returns.

* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawRemoteToPDF):

Canonical link: <a href="https://commits.webkit.org/290300@main">https://commits.webkit.org/290300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/caf8287a58618ae10a5fb5b4e3d2fdbe0893ee32

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89547 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9074 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94539 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40314 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9463 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17353 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68977 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26631 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92548 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7264 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81265 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49342 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6995 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35649 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-grid/grid-definition/grid-auto-fill-columns-001.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39420 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77349 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36650 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96367 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16729 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77848 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16984 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77066 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77164 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19054 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21597 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20171 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9896 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16742 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22055 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16483 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18265 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -194,17 +197,23 @@ ImageBufferBackend* RemoteImageBufferProxy::ensureBackend() const
+    if (UNLIKELY(!remoteRenderingBackendProxy))
+        return nullptr;
+
+    RELEASE_LOG(RemoteLayerBuffers, "[renderingBackend=%" PRIu64 "] RemoteImageBufferProxy::ensureBackendCreated - waitForAndDispatchImmediately returned error: %" PUBLIC_LOG_STRING,
```
Might be worth continuing to log these errors, just with a 0 identifier if the backend proxy has gone away.

shallawa replied:
Fixed.

shallawa committed:
[Site isolation] Record the page snapshotting to a DisplayList before sinking it to a PDFDocument
https://bugs.webkit.org/show_bug.cgi?id=285726#
rdar://142664977

Reviewed by Matt Woodrow.

Introduce ImageBufferDisplayListBackend and ImageBufferRemoteDisplayListBackend
to support recording the snapshot to a DisplayList.

Make drawRemoteToPDF() creates a ImageBufferRemoteDisplayListBackend. Its GPUProcess
twin ImageBufferDisplayListBackend will record the drawing commands into a
DisplayList.

Make didDrawRemoteToPDF() calls sinkIntoPDFDocument() on the main thread because
DisplayList::Replayer needs a ControlFactory and ControlFactory::shared() has to
be called on the main thread. Then send the result SharedBuffer to the UIProcess.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::sinkIntoPDFDocument):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferDisplayListBackend.cpp: Added.
(WebCore::ImageBufferDisplayListBackend::create):
(WebCore::ImageBufferDisplayListBackend::ImageBufferDisplayListBackend):
(WebCore::ImageBufferDisplayListBackend::context):
(WebCore::ImageBufferDisplayListBackend::copyNativeImage):
(WebCore::ImageBufferDisplayListBackend::sinkIntoPDFDocument):
(WebCore::ImageBufferDisplayListBackend::debugDescription const):
* Source/WebCore/platform/graphics/ImageBufferDisplayListBackend.h: Copied from Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp.
(WebCore::ImageBufferDisplayListBackend::calculateMemoryCost):
* Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp:
(WebCore::DisplayList::DrawingContext::replayDisplayList):
The PDFDocument has to be generated on the main thread because the main frame
DisplayList will eventually reference other DisplayLists which were created by
different WorkQueues. So clearing the DisplayList, once it is replayed back, will
delete the resources in the ResourceHeap on the main thread. This will fire the
assertion: "Unsafe to ref/deref from different threads". So keep the DisplayList
items till they are deleted by the destructor of DisplayList::DrawingContext.

* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didDrawRemoteToPDF):
(WebKit::allocateImageBufferInternal):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemoteDisplayListBackend.cpp: Added.
(WebKit::ImageBufferRemoteDisplayListBackend::create):
(WebKit::ImageBufferRemoteDisplayListBackend::ImageBufferRemoteDisplayListBackend):
(WebKit::ImageBufferRemoteDisplayListBackend::~ImageBufferRemoteDisplayListBackend):
(WebKit::ImageBufferRemoteDisplayListBackend::createNativeImageReference):
(WebKit::ImageBufferRemoteDisplayListBackend::debugDescription const):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferRemoteDisplayListBackend.h: Copied from Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):

(WebKit::RemoteImageBufferProxy::ensureBackend const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::sendSync):
Fix unrelated mac-safer-cpp errors. Make these functions make early returns.

* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawRemoteToPDF):

Canonical link: https://commits.webkit.org/290300@main
